File: scfw/verifiers/age_verifier/pypi.py
"""
Provides utilities for querying the PyPI registry to discover package creation dates.
"""

from datetime import datetime
from dateutil import parser as datetime_parser

import requests


def get_creation_datetime_utc(package_name: str) -> datetime:
    """
    Return the creation date for the given PyPI package.

    Args:
        package_name: The `str` package name whose creation date should be queried.

    Returns:
        A UTC `datetime` representing the creation date of `package_name` on PyPI.

    Raises:
        requests.HTTPError: Failed to query the PyPI registry.
        requests.exceptions.JSONDecodeError: Failed to parse query response as JSON.
        RuntimeError:
            * Package metadata missing required fields.
            * No upload timestamps found in package metadata.
        dateutil.ParserError: Failed to parse publication datetime.
    """
    r = requests.get(f"https://pypi.org/pypi/{package_name}/json")
    r.raise_for_status()

    package_metadata = r.json()
    releases = package_metadata.get("releases")
    if not releases:
        raise RuntimeError("Package metadata missing required fields")

    upload_datetimes = set()
    for _, metadatas in releases.items():
        for metadata in metadatas:
            upload_timestamp = metadata.get("upload_time_iso_8601")
            if not upload_timestamp:
                continue

            upload_datetimes.add(datetime_parser.parse(upload_timestamp))

    if not upload_datetimes:
        raise RuntimeError("No upload timestamps found in package metadata")

    return sorted(upload_datetimes)[0]

File: internal/ecosystem/pypi.go
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024-present Datadog, Inc.

package ecosystem

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"slices"
	"time"
)

// pyPIRegistryDomains are the URL domains associated with the PyPI registry, ported from
// ecosystem.py's ECOSYSTEM.registry_domains in the Python reference.
var pyPIRegistryDomains = []string{"pypi.org", "files.pythonhosted.org"}

// isPyPIRegistrySource reports whether source is a known PyPI registry URL, ported from
// package.py's Package.has_registry_source in the Python reference.
func isPyPIRegistrySource(source string) bool {
	u, err := url.Parse(source)
	if err != nil {
		return false
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return false
	}
	return slices.Contains(pyPIRegistryDomains, u.Hostname())
}

// resolvePyPIPublishDate returns the publication date of the given package version on
// PyPI, taken to be the latest of the release's upload timestamps.
//
// A non-empty source that isn't a known PyPI registry URL has no publish date to resolve;
// this is a no-op rather than an error, leaving it to the caller to decide how to handle
// an unresolved publish date, and the zero value of time.Time is returned in that case.
// An empty source is assumed to be the PyPI registry.
func resolvePyPIPublishDate(ctx context.Context, name, version, source string) (time.Time, error) {
	if source != "" && !isPyPIRegistrySource(source) {
		return time.Time{}, nil
	}

	url := fmt.Sprintf("https://pypi.org/pypi/%s/json", name)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return time.Time{}, fmt.Errorf("failed to build PyPI request: %w", err)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return time.Time{}, fmt.Errorf("failed to query PyPI registry: %w", err)
	}
	defer resp.Body.Close() //nolint:errcheck

	if resp.StatusCode >= 300 {
		return time.Time{}, fmt.Errorf("received unexpected status from PyPI registry: %s", resp.Status)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return time.Time{}, fmt.Errorf("failed to read PyPI response: %w", err)
	}

	return parsePyPIReleaseMetadata(body, name, version)
}

// parsePyPIReleaseMetadata decodes the JSON returned by PyPI's package metadata endpoint
// and returns the latest upload timestamp recorded for the given package version.
func parsePyPIReleaseMetadata(body []byte, name, version string) (time.Time, error) {
	var metadata struct {
		Releases map[string][]struct {
			UploadTimeISO8601 string `json:"upload_time_iso_8601"`
		} `json:"releases"`
	}
	if err := json.Unmarshal(body, &metadata); err != nil {
		return time.Time{}, fmt.Errorf("failed to parse PyPI response as JSON: %w", err)
	}

	releases, ok := metadata.Releases[version]
	if !ok || len(releases) == 0 {
		return time.Time{}, fmt.Errorf("package metadata missing required fields")
	}

	var latest time.Time
	for _, release := range releases {
		if release.UploadTimeISO8601 == "" {
			continue
		}
		t, err := time.Parse(time.RFC3339Nano, release.UploadTimeISO8601)
		if err != nil {
			return time.Time{}, fmt.Errorf("failed to parse publication datetime %q: %w", release.UploadTimeISO8601, err)
		}
		if t.After(latest) {
			latest = t
		}
	}

	if latest.IsZero() {
		return time.Time{}, fmt.Errorf("no publication timestamp for version %s of package %s", version, name)
	}

	return latest, nil
}
